File: Python/kmeanspp.py
"""
Usage:
    kmeanspp.py generate_dataset <dataset_size>
    kmeanspp.py clusterize <dataset> <number_of_clusters>
    kmeanspp.py export
"""

import sys
import csv
import math
import random
from copy import deepcopy
from docopt import docopt

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.cluster = None

    distanceTo = lambda self, p : math.sqrt(math.pow(self.x - p.x, 2) + math.pow(self.y - p.y, 2))

class Cluster:
    def __init__(self, center):
        self.center = center
        self.points = []
        
class Dataset:
    def __init__(self, points):
        self.points = points
        self.clusters = []

    def initializeCenters(self, n):
        init_centers = [random.choice(self.points)]
        current_center = init_centers[0]
        
        for i in range(1, n):
            candidate_points = self.points[:]
            for p in init_centers:
                for c in candidate_points:
                    if p.x == c.x and p.y == c.y:
                        candidate_points.remove(c)
           
            weight_dividor = sum(math.pow(p.distanceTo(current_center), 2) for p in candidate_points)
            max_weight = 0
            best_candidate = None
            for p in candidate_points:
                weight = math.pow(p.distanceTo(current_center), 2) / weight_dividor
                if weight > max_weight:
                    max_weight = weight
                    best_candidate = p
            
            current_center = best_candidate
            init_centers.append(current_center)
        
        for c in init_centers:
            self.clusters.append(Cluster(c))

        for p in self.points:
            self.clusters[0].points.append(p)
            p.cluster = self.clusters[0]

    def updateClusterMembers(self):
        for c in self.clusters:
            for p in self.points:
                if p.distanceTo(c.center) < p.distanceTo(p.cluster.center):
                    p.cluster.points.remove(p)
                    c.points.append(p)
                    p.cluster = c
            c.center = barycenterFromList(c.points)

def datasetFromCSV(filename):
    points =[]
    with open(filename, 'r') as f:
        reader = csv.reader(f, delimiter=',')
        for e in reader:
            if not any(p.x == int(e[0]) and p.y == int(e[1]) for p in points):
                points.append(Point(int(e[0]), int(e[1])))

    return Dataset(points)

barycenterFromList = lambda points : Point(sum(p.x for p in points) / len(points), sum(p.y for p in points) / len(points)) 

if __name__ == '__main__':
    args = docopt(__doc__)

    if args["generate_dataset"]:
        print("[ ] Generating " + args["<dataset_size>"] + " 2D points into dataset.csv...")
        with(open("dataset.csv", 'w')) as out:
            w = csv.writer(out)
            for i in range(int(args["<dataset_size>"])):
                w.writerow([random.randint(0, 1000), random.randint(0, 1000)])

        print("[+] Done.")

    elif args["clusterize"]:
        print("[ ] Loading dataset...")
        dataset = datasetFromCSV(args["<dataset>"])
        print("[ ] Choosing initial cluster centers...")
        dataset.initializeCenters(int(args["<number_of_clusters>"]))
        
        centers = list(c.center for c in dataset.clusters)
        previous_centers = [Point(-1, -1)]
        i = 0
        
        while(any(c.x != p.x or c.y != p.y for (c, p) in zip(centers, previous_centers))):
            print("[ ] Updating clusters (iteration " + str(i) + ")...")
            previous_centers = deepcopy(centers)
            dataset.updateClusterMembers()
            centers = list(c.center for c in dataset.clusters)
            i = i+1
   
        print("[ ] Clusters found, writing to out.csv...")
        with(open("out.csv", 'w', newline='')) as out:
            w = csv.writer(out)
            for i in range(len(dataset.clusters)):
                for p in dataset.clusters[i].points:
                    w.writerow([p.x, p.y, str(i)])

        print("[+] Done.")

    elif args["export"]:
        print("[ ] Exporting points from out.csv to out.svg...")
        svg_str = "<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='1000' height='1000'>\n"
        svg_colors = ["black", "blue", "red", "green", "yellow", "gray", "gold", "tan", "olivedrab", "aqua"]

        with open("out.csv", 'r') as f:
            reader = csv.reader(f, delimiter=',')
            for e in reader:
                svg_str = svg_str + "   <circle cx='" + e[0] + "' cy='" + e[1] + "' r='5' fill='" + svg_colors[int(e[2]) % len(svg_colors)] + "' />\n"

        svg_str = svg_str + "</svg>\n"
        with open("out.svg", 'w') as out:
            out.write(svg_str)

        print("[+] Done.")

File: Java/src/Kmeanspp.java
import java.io.FileWriter;
import java.io.FileReader;
import java.io.BufferedReader;
import java.util.Random;
import java.util.ArrayList;
import java.lang.Math;

public class Kmeanspp {

    enum Op {
        generate,
        clusterize,
        export
    }

    Op op;
    int dsize, clusters;
    String dfile;

    public Kmeanspp() {
        op = Op.export;
    }

    public Kmeanspp(int dsize) {
        op = Op.generate;
        this.dsize = dsize;
    }

    public Kmeanspp(String dfile, int clusters) {
        op = Op.clusterize;
        this.clusters = clusters;
        this.dfile = dfile;
    }

    public void run() {
        switch(op) {
            case generate:
                generate_dataset();
                break;

            case clusterize:
                clusterize();
                break;

            case export:
                export("out.csv", "out.svg");
                break;
        }
    }

    private void generate_dataset() {
        System.out.println("[ ] Generating " + dsize + " 2D points into dataset.csv...");

        try(FileWriter fw = new FileWriter("dataset.csv")) {
            Random rand = new Random();

            for(int i = 0; i < dsize; i++)
                fw.write(rand.nextInt(1000) + "," + rand.nextInt(1000) + "\n");

            fw.close();
            System.out.println("[+] Done.");
        }

        catch(Exception e) {
            System.out.println("[!] File error");
        }
    }

    private void clusterize() {
        try {
            System.out.println("[ ] Loading dataset...");
            Dataset data = new Dataset(dfile, clusters);

            System.out.println("[ ] Choosing initial cluster centers...");
            data.initializeCenters();

            data.updateClustersLoop();

            System.out.println("[ ] Clusters found, writing to out.csv...");
            FileWriter fw = new FileWriter("out.csv");

            for(Point p : data.points)
                fw.write(p.toString() + "\n");

            fw.close();
            System.out.println("[+] Done.");
        }

        catch(Exception e) {
            e.printStackTrace();
        }
    }

    // Write svg from clusterized csv
    private void export(String ffrom, String fto) {
        System.out.println("[ ] Exporting points from out.csv to out.svg...");

        try(BufferedReader br = new BufferedReader(new FileReader(ffrom))) {

            String svgStr = "<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='1000' height='1000'>\n";

            for(String line; (line = br.readLine()) != null;) {
                String[] l = line.trim().split(",");
                svgStr += "    <circle cx='" + l[0] + "' cy='" + l[1] + "' r='5' fill='"
                       + Dataset.Colors[Integer.parseInt(l[2]) % Dataset.Colors.length] + "' />\n";
            }

            br.close();
            svgStr += "</svg>\n";

            FileWriter fw = new FileWriter(fto);
            fw.write(svgStr);

            fw.close();
            System.out.println("[+] Done");
        }

        catch(Exception e) {
            System.out.println("[!] File error");
            e.printStackTrace();
        }
    }
}

class Point {

    Cluster cluster;
    double x, y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double squaredDistanceTo(Point p) {
        return Math.pow(x - p.x, 2) + Math.pow(y - p.y, 2);
    }

    public double distanceTo(Point p) {
        return Math.sqrt(squaredDistanceTo(p));
    }

    public boolean equals(Point p) {
        return x == p.x && y == p.y;
    }

    public String toString() {
        return (int) x + "," + (int) y + (cluster == null ? "" : "," + cluster.id);
    }
}

class Cluster {

    ArrayList<Point> points;
    Point center;
    int id;

    public Cluster(Point center, int id) {
        this.center = center;
        this.id = id;
        points = new ArrayList<Point>();
    }

    Point barycenter() {
        double xs = 0;
        double ys = 0;
        int s = points.size();

        for(Point p: points) {
            xs += p.x;
            ys += p.y;
        }

        return new Point(xs / s, ys / s);
    }
}

class Dataset {

    int nclust;
    ArrayList<Point> points;
    ArrayList<Cluster> clusters;

    static String[] Colors = {"black", "blue", "red", "green", "yellow", "gray", "gold", "tan", "olivedrab", "aqua"};

    // Read dataset from csv
    public Dataset(String file, int nclust) throws Exception {

        this.nclust = nclust;
        points = new ArrayList<Point>();
        clusters = new ArrayList<Cluster>(nclust);

        try(BufferedReader br = new BufferedReader(new FileReader(file))) {

            for(String line; (line = br.readLine()) != null;) {
                String[] l = line.trim().split(",");
                points.add(
                    new Point(
                        Integer.parseInt(l[0]),
                        Integer.parseInt(l[1])
                    )
                );
            }

            br.close();

            if(nclust > points.size()) {
                System.out.println("[!] Error: we are looking for more clusters than there are points in the dataset");
                throw new IllegalArgumentException();
            }
        }

        catch(java.io.FileNotFoundException e) {
            System.out.println("[!] File error");
        }
    }

    // Choose the initial cluster centers
    void initializeCenters() {

        Random rand = new Random();
        ArrayList<Point> initCenters = new ArrayList<Point>(nclust);

        initCenters.add(points.get(rand.nextInt(points.size())));
        Point currentCenter = initCenters.get(0);

        for(int i=1; i<nclust; i++) {
            ArrayList<Point> candidatePoints = new ArrayList<Point>(points);
            ArrayList<Point> toRemove = new ArrayList<Point>();

            for(Point p : initCenters)
                for(Point c : candidatePoints)
                    if(c.equals(p))
                        toRemove.add(p);

            candidatePoints.removeAll(toRemove);

            double weightDividor = 0;

            for(Point p : candidatePoints)
                weightDividor += p.squaredDistanceTo(currentCenter);

            double maxWeight = 0;
            Point bestCandidate = null;

            for(Point p : candidatePoints) {
                double weight = p.squaredDistanceTo(currentCenter) / weightDividor;
                if(weight > maxWeight) {
                    maxWeight = weight;
                    bestCandidate = p;
                }
            }

            currentCenter = bestCandidate;
            initCenters.add(currentCenter);
        }

        // Create the clusters
        for(int i=0; i<initCenters.size(); i++)
            clusters.add(new Cluster(initCenters.get(i), i));

        // Affect every point to the first cluster
        for(Point p : points) {
            clusters.get(0).points.add(p);
            p.cluster = clusters.get(0);
        }
    }

    boolean updateClusters() {

        boolean nomodif = true;
        for(Cluster c : clusters) {

            if(c.points.size() > 0)
                c.center = c.barycenter();

            ArrayList<Point> toUpdate = new ArrayList<Point>();

            for(Point p : points) {
                if(p.distanceTo(p.cluster.center) > p.distanceTo(c.center)) {
                    p.cluster = c;

                    if(!toUpdate.contains(p))
                        toUpdate.add(p);

                    nomodif = false;
                }
            }

            for(Point p : toUpdate) {
                c.points.remove(p);
                p.cluster.points.add(p);
            }
        }

        return nomodif;
    }

    // Keep recomputing the clusters until we get no change by doing so
    void updateClustersLoop() {
        int i = 0;

        do {
            System.out.println("[ ] Updating clusters (iteration " + i + ")...");
            i++;
        } while(!updateClusters());
    }
}
